aten/src/ATen/cpu/vec256/vec256_qint.h

```diff
@@ -154,6 +154,223 @@ inline void __attribute__((always_inline)) QuantizeAvx2(
+int64_t custom_mm256_hsum_epi64_ignore_overflow(__m256i x) {
+  // ( -, -, y1, y0)
+  const __m256i loDual = x;
+  // ( -, -, y3, y2)
+  const __m256i hiDual =
+    _mm256_permute2f128_si256(x, x, _MM_SHUFFLE(0, 1, 0, 1));
+  // ( -, -, y1 + y3, y0 + y2)
+  const __m256i sumDual = _mm256_add_epi64(loDual, hiDual);
+  // ( -, -, -, y0 + y2)
+  const __m256i lo = sumDual;
+  // ( -, -, -, y1 + y3)
+  const __m256i hi = _mm256_permute4x64_epi64(sumDual, _MM_SHUFFLE(1, 1, 1, 1));
+  // ( -, -, -, y1 + y3 + y0 + y2)
+  const __m256i sum = _mm256_add_epi64(lo, hi);
+  const int64_t res = _mm256_extract_epi64(sum, 0);
+  return res;
+}
+
+// horizontal sums signed i32
+// x = (x7, x6, x5, x4, x3, x2, x1, x0)
+int64_t custom_mm256_hsum_epi32(__m256i x) {
+  // upsample to int64 to avoid overflow
+  // (x7, x6, x5, x4), int32
+  const __m128i xHalf1 = _mm256_extractf128_si256(x, 1);
+  // (x3, x2, x1, x0), int32
+  const __m128i xHalf2 = _mm256_castsi256_si128(x);
+  // (x7, x6, x5, x4), int64
+  const __m256i xHalf1_64 = _mm256_cvtepi32_epi64(xHalf1);
+  // (x3, x2, x1, x0), int64
+  const __m256i xHalf2_64 = _mm256_cvtepi32_epi64(xHalf2);
+  // (x7 + x3, x6 + x2, x5 + x1, x4 + x0) == (y3, y2, y1, y0)
+  const __m256i xHalfSum = _mm256_add_epi64(xHalf1_64, xHalf2_64);
+  return custom_mm256_hsum_epi64_ignore_overflow(xHalfSum);
+}
+
+// horizontal sums signed i16
+// x = (x15, ..., x0)
+int64_t custom_mm256_hsum_epi16(__m256i x) {
+  // (x15, ..., x8), int16
+  const __m128i xHalf1 = _mm256_extractf128_si256(x, 1);
+  // (x7, ..., x0), int16
+  const __m128i xHalf2 = _mm256_castsi256_si128(x);
+  // (x15, ..., x8), int32
+  const __m256i xHalf1_32 = _mm256_cvtepi16_epi32(xHalf1);
+  // (x7, ..., x0), int32
+  const __m256i xHalf2_32 = _mm256_cvtepi16_epi32(xHalf2);
+  // (x15 + x7, ..., x8 + x0) = (y7, ..., y0)
+  const __m256i xSum = _mm256_add_epi32(xHalf1_32, xHalf2_32);
+  // ( -, -, -, -, y3, y2, y1, y0)
+  const __m256i xHalfLo = xSum;
+  // ( -, -, -, -, y7, y6, y5, y4)
+  const __m256i xHalfHi = _mm256_permute2f128_si256(xSum, xSum, _MM_SHUFFLE(0, 1, 0, 1));
+  // ( -, -, -, -, y3 + y7, y2 + y6, y1 + y5, y0 + y4)
+  const __m256i xHalfSum = _mm256_add_epi32(xHalfLo, xHalfHi);
+  // ( -, -, -, -, -, -, y1 + y5, y0 + y4)
+  const __m256i xQuadLo = xHalfSum;
+  // ( -, -, -, -, -, -, y3 + y7, y2 + y6)
+  const __m256i xQuadHi = _mm256_permute4x64_epi64(xHalfSum, _MM_SHUFFLE(1, 1, 1, 1));
+  // ( -, -, -, -, -, -, y1 + y5 + y3 + y7, y0 + y4 + y2 + y6)
+  const __m256i sumQuad = _mm256_add_epi32(xQuadLo, xQuadHi);
+  // ( -, -, -, -, -, -, -, y0 + y4 + y2 + y6)
+  const __m256i lo = sumQuad;
+  // ( -, -, -, -, -, -, -, y1 + y5 + y3 + y7)
+  const __m256i hi = _mm256_shuffle_epi32(sumQuad, _MM_SHUFFLE(1, 1, 1, 1));
+  const __m256i sum = _mm256_add_epi32(lo, hi);
+  int32_t res = _mm256_extract_epi32(sum, 0);
+  return static_cast<int64_t>(res);
+}
+
+// horizontal sums signed i8
+// x = (x31, ..., x0)
+int custom_mm256_hsum_epi8(__m256i x) {
```

Comment:
The return value is immediately getting widened to int64_t at the callsite. Should we make that explicit in the return type here (and below?)